[63888] Move project selector to the sidebar

bsatarnejad opened the pull request:
# Ticket
https://community.openproject.org/wp/63888

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What are you trying to accomplish?
Remove the project selector from top menu left, and move it to side bar menu. 

## Screenshots
![Screenshot 2025-05-21 at 11 55 09](https://github.com/user-attachments/assets/88a92b4d-15f1-4dcd-b9de-a4b62b31a041)

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

bsatarnejad committed:
Remove project selector from top menu left

Move the project selector to side bar menu

It doesn't show project selector in log in page any more, so we don't need to test showing favorite projects

Fix spacing in sidebar menu

Use a smaller header title for projects list modal

remove op-app-menu classes

Show project name on a tooltip and make the text ellipsis in menu

Add a divider between project selector and menu list

Remove border radius

Use spot divider

Undo changes for spec

Fix test by adding view_project permission to anonymous user

Use hr tag instead of spot divider

Add some space around the project selector

Fix spacing around the divider and fix the hover effect width

Reduce spacing to the top to match bottom spacing

remove overflow y for main menu sidebar and use button instead of anchor tag for trigger element

reduce the height of project selector from menu sidebar height, add a scroll for the menu when there is no sub menu

HDinger commented on app/components/open_project/common/submenu_component.sass:
```diff
@@ -53,6 +53,7 @@
       background: var(--main-menu-bg-hover-background)
       color: var(--main-menu-hover-font-color)
       border-color: var(--main-menu-hover-border-color)
+      border-radius: var(--borderRadius-medium)
```
This looks weird imho 🙈 but since it was specified in the AC it is probably okay 🤷

HDinger commented on spec/features/projects/favorite_spec.rb:
```diff
@@ -163,19 +163,4 @@
     end
   end
-
-  context "as an Anonymous User with not login required", with_settings: { login_required: false } do
```
Why did you delete this spec? It is still valid...

HDinger commented on frontend/src/global_styles/common/header/app-menu.sass:
```diff
@@ -26,7 +26,7 @@
     border-top: 0
     border-bottom: var(--header-border-bottom-width) solid transparent
-    border-radius: 0
+    border-radius: var(--borderRadius-medium)
```
Are you sure, that the top menu items should be affected by that change as well. It feels again weird.
<img width="50" alt="Bildschirmfoto 2025-06-03 um 11 40 07" src="https://github.com/user-attachments/assets/4da4fe0d-e5b6-4fc1-9e2e-832bfa3f78b3" />

bsatarnejad replied:
![Screenshot 2025-06-04 at 07 44 40](https://github.com/user-attachments/assets/609183b9-c381-46d0-9983-76476826720c)

Yes, The figma design shows something like this.

HDinger replied:
Well, this is not how it looks with this PR either. Additionally, this is a futuristic design. This will not come in the scope of this ticket. In the other designs, the top bar is untouched (which is also part of the AC)

HDinger commented on frontend/src/global_styles/common/header/app-menu.sass:
```diff
@@ -36,7 +36,7 @@
     text-decoration: none
     min-width: 0px
-    padding: 0 var(--main-menu-x-spacing)
+    padding: 0 var(--base-size-8)
```
What is the benefit of that change? `--main-menu-x-spacing` is a variable we use in multiple places to ensure that all the items are aligned on the same line. With your change, the items will not be aligned any more when we change the spacing.

HDinger commented on frontend/src/global_styles/layout/_main_menu.sass:
```diff
@@ -119,7 +130,7 @@ $arrow-left-width: 40px
     font-size: var(--main-menu-font-size)
     font-style: normal
-    padding: 0 var(--main-menu-x-spacing)
+    padding: 0 var(--base-size-8)
```
Same as before

HDinger commented on frontend/src/global_styles/layout/_main_menu.sass:
```diff
@@ -54,6 +64,7 @@ $arrow-left-width: 40px
     +allow-vertical-scrolling
     height: calc(100vh - var(--header-height))
     position: relative
+    padding: 0 var(--base-size-8)
```
This shrinks the whole available space for the sidebar, although you probably only want it for the project selector. Further, the selector and the items are not aligned because of this.

<img width="234" alt="Bildschirmfoto 2025-06-03 um 11 52 21" src="https://github.com/user-attachments/assets/394e7922-5ebe-41e6-a333-97877f5aa853" />

bsatarnejad replied:
![Screenshot 2025-06-04 at 11 16 00](https://github.com/user-attachments/assets/6c3c6cd9-e820-4c34-8717-f170ef31d230)

There is a padding for the whole sidebar.

HDinger commented on frontend/src/global_styles/layout/_main_menu.sass:
```diff
@@ -165,6 +176,14 @@ $arrow-left-width: 40px
+      @include main-item-wrapper-border
+
+      & ~ .toggler
+        @include main-item-wrapper-toggler-border
```
We have a [similar block above](https://github.com/opf/openproject/blob/dev/frontend/src/global_styles/layout/_main_menu.sass#L146-L159), why did you not put this code there? Further, what effect is that supposed to have? I see no difference when removing that..

HDinger reviewed (changes requested):
I think there are some styling refinements to be done:

- [x] The line between the project selector and the menu is missing
- [x] The hover effect of the project selector is way to big compared to the other items
- [x] The alignment between the project selector and the second-level items is off
- [x] When I have a very long title, it is currently truncated and I have no chance to know the rest of the title. I guess, a quick fix would be to show a tooltip on hover. It may be an option to wrap instead of truncate but that would need a designers decicion.
- [x] The project selector still has a lot of the `op-app-menu` classes. Thus, it takes for example the hover color of the header instead of the main menu. Ideally, we'd get rid of these classes inside the project selector.

HDinger commented on app/views/layouts/base.html.erb:
```diff
@@ -103,6 +103,8 @@ See COPYRIGHT and LICENSE files for more details.
         <opce-main-menu-resizer></opce-main-menu-resizer>
         <div id="menu-sidebar">
+          <%= render_projects_top_menu_node %>
+          <div class="spot-divider"></div>
```
I am not a big fan of seeing the spot classes being used here again. It implements no accessibility features and it does not adapt to the sidebar colours. Maybe something like this could work:

```
         <%= render(Primer::BaseComponent.new(tag: :hr,
                                              classes: "whatever",
                                              role: :presentation,
                                              aria: {
                                                hidden: true
                                              })) %>
```

and then style it to respect the colors of the sidebar

HDinger reviewed (changes requested):
- [ ] In the AC it is written, that the drop down should have some spacing to the left side. Currently, it is glued to the left side
- [x] For me it looks like there is a margin missing around the project selector. It looks somehow squashed compared to how much space for example the submenu headers have. In figma, it also looks as if there was some margin around it. Maybe you have to double check with @marcalcobe

bsatarnejad commented:
> * The hover effect of the project selector is way to big compared to the other items

I talked with Marc about the spacing on the left side of the drop modal, the decision is to ignore it now.

bsatarnejad commented:
> * [ ]  In the AC it is written, that the drop down should have some spacing to the left side. Currently, it is glued to the left side
> * [ ]  For me it looks like there is a margin missing around the project selector. It looks somehow squashed compared to how much space for example the submenu headers have. In figma, it also looks as if there was some margin around it. Maybe you have to double check with @marcalcobe

I talked with Marc about the spacing on the left side of the drop modal, the decision is to ignore it now.

HDinger reviewed (commented):
The only thing left for me is the hover effect. In figma it is not really specified, but the way it currently looks, feels weird:

<img width="236" alt="Bildschirmfoto 2025-06-11 um 13 28 53" src="https://github.com/user-attachments/assets/c20537d6-03bc-4a14-ba09-8b6adb9952c1" />

I guess it needs to be either full width or have a spacing to left as well. Please contact the UX team for that.

HDinger commented on frontend/src/app/shared/components/header-project-select/header-project-select.component.html:
```diff
@@ -5,22 +5,26 @@
   alignment="bottom-start"
   class="op-project-list-modal"
 >
-  <button
```
This change results in the project selector not being reachable via keyboard anymore as the anchor does not have an `href`. Semantically, `button` is also the correct tag..

HDinger commented:
One other thing that just came to my mind: I guess, the project selector should be sticky and not scroll with the other elements when the menu is rather long

HDinger reviewed (approved):
Code changes look good 👍 As discussed, we will keep that open until the hamburger icon is moved as well

HDinger committed:
Move toggleButton for menu into the sidebar

HDinger committed:
Remove main-menu-bg-hover-color from set of configurable design variables. Instead we use the Primer behaviour for ActionLists

HDinger committed:
Take care that the main menu resizer toggles when the menu is collapsed/expanded

Clean-up sass code

Avoid that the type selector in the WP from is scrolled two far when being activated. Currently, it is scrolled twice (one time when activating it and once again when selecting an item). Thus the dropdown overlaps the actual input

Move logo to the left side

Make naming consistent for MainMenuToggleComponent and add a component test

Let menu toggle button scroll with the rest of the page

Move modules menu to the left and show it on mobile

HDinger committed:
Try whether that stabilizes the test

oliverguenther committed:
Don't cache only on user

The user itself doesn't update when a new attribute help text or role is
added to them

HDinger committed:
Hide menu button on pages where there is no sidebar

HDinger committed:
Avoid heavy scrolling after already activating the field to avoid that the autocompleter dropdown overlaps the input

HDinger committed:
Merge pull request #19170 from opf/implementation/63889-move-toggle-button-to-the-sidebar

[63889] Move toggle button to the sidebar

HDinger committed:
Remove unnecessary log statements

[ci skip]
